<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.269">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="dcterms.date" content="2023-08-02">

<title>Giotto Suite with Xenium Data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="xenium_MOB_files/libs/clipboard/clipboard.min.js"></script>
<script src="xenium_MOB_files/libs/quarto-html/quarto.js"></script>
<script src="xenium_MOB_files/libs/quarto-html/popper.min.js"></script>
<script src="xenium_MOB_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="xenium_MOB_files/libs/quarto-html/anchor.min.js"></script>
<link href="xenium_MOB_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="xenium_MOB_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="xenium_MOB_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="xenium_MOB_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="xenium_MOB_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#data-and-package-loading" id="toc-data-and-package-loading" class="nav-link active" data-scroll-target="#data-and-package-loading"><span class="toc-section-number">1</span>  Data and Package Loading</a></li>
  <li><a href="#import-segmentation-data" id="toc-import-segmentation-data" class="nav-link" data-scroll-target="#import-segmentation-data"><span class="toc-section-number">2</span>  Import Segmentation Data</a></li>
  <li><a href="#create-a-giotto-object" id="toc-create-a-giotto-object" class="nav-link" data-scroll-target="#create-a-giotto-object"><span class="toc-section-number">3</span>  Create a Giotto Object</a>
  <ul class="collapse">
  <li><a href="#create-a-giotto-object-with-convenience-function" id="toc-create-a-giotto-object-with-convenience-function" class="nav-link" data-scroll-target="#create-a-giotto-object-with-convenience-function"><span class="toc-section-number">3.1</span>  Create a Giotto Object with Convenience Function</a></li>
  </ul></li>
  <li><a href="#generate-aggregated-expression-based-on-feature-and-boundary-polygon-information" id="toc-generate-aggregated-expression-based-on-feature-and-boundary-polygon-information" class="nav-link" data-scroll-target="#generate-aggregated-expression-based-on-feature-and-boundary-polygon-information"><span class="toc-section-number">4</span>  Generate Aggregated Expression based on Feature and Boundary (polygon) Information</a>
  <ul class="collapse">
  <li><a href="#assign-polygon-overlaps-information-to-expression-matrix" id="toc-assign-polygon-overlaps-information-to-expression-matrix" class="nav-link" data-scroll-target="#assign-polygon-overlaps-information-to-expression-matrix"><span class="toc-section-number">4.1</span>  Assign polygon overlaps information to expression matrix</a></li>
  <li><a href="#add-metadata" id="toc-add-metadata" class="nav-link" data-scroll-target="#add-metadata"><span class="toc-section-number">4.2</span>  Add Metadata</a></li>
  <li><a href="#data-filtering" id="toc-data-filtering" class="nav-link" data-scroll-target="#data-filtering"><span class="toc-section-number">4.3</span>  Data filtering</a></li>
  <li><a href="#add-data-statistics" id="toc-add-data-statistics" class="nav-link" data-scroll-target="#add-data-statistics"><span class="toc-section-number">4.4</span>  Add data statistics</a></li>
  <li><a href="#normalize-expression" id="toc-normalize-expression" class="nav-link" data-scroll-target="#normalize-expression"><span class="toc-section-number">4.5</span>  Normalize expression</a></li>
  <li><a href="#calculate-highly-variable-features" id="toc-calculate-highly-variable-features" class="nav-link" data-scroll-target="#calculate-highly-variable-features"><span class="toc-section-number">4.6</span>  Calculate highly variable features</a></li>
  </ul></li>
  <li><a href="#dimension-reduction-and-clustering" id="toc-dimension-reduction-and-clustering" class="nav-link" data-scroll-target="#dimension-reduction-and-clustering"><span class="toc-section-number">5</span>  Dimension reduction and clustering</a>
  <ul class="collapse">
  <li><a href="#pca" id="toc-pca" class="nav-link" data-scroll-target="#pca"><span class="toc-section-number">5.1</span>  PCA</a></li>
  <li><a href="#tsne-and-umap" id="toc-tsne-and-umap" class="nav-link" data-scroll-target="#tsne-and-umap"><span class="toc-section-number">5.2</span>  tSNE and UMAP</a></li>
  <li><a href="#snn-and-leiden-clustering" id="toc-snn-and-leiden-clustering" class="nav-link" data-scroll-target="#snn-and-leiden-clustering"><span class="toc-section-number">5.3</span>  sNN and Leiden clustering</a></li>
  </ul></li>
  <li><a href="#spatial-expression-patterns" id="toc-spatial-expression-patterns" class="nav-link" data-scroll-target="#spatial-expression-patterns"><span class="toc-section-number">6</span>  Spatial expression patterns</a></li>
  <li><a href="#view-images" id="toc-view-images" class="nav-link" data-scroll-target="#view-images"><span class="toc-section-number">7</span>  View images</a></li>
  <li><a href="#subsetting-the-giotto-object" id="toc-subsetting-the-giotto-object" class="nav-link" data-scroll-target="#subsetting-the-giotto-object"><span class="toc-section-number">8</span>  Subsetting the Giotto Object</a></li>
  <li><a href="#custom-segmentations" id="toc-custom-segmentations" class="nav-link" data-scroll-target="#custom-segmentations"><span class="toc-section-number">9</span>  Custom Segmentation(s)</a>
  <ul class="collapse">
  <li><a href="#cell-typing-for-each-segmentation-method-by-page-enrichment" id="toc-cell-typing-for-each-segmentation-method-by-page-enrichment" class="nav-link" data-scroll-target="#cell-typing-for-each-segmentation-method-by-page-enrichment"><span class="toc-section-number">9.1</span>  Cell Typing for each Segmentation Method by PAGE Enrichment</a></li>
  <li><a href="#plot-cell-type-distributions" id="toc-plot-cell-type-distributions" class="nav-link" data-scroll-target="#plot-cell-type-distributions"><span class="toc-section-number">9.2</span>  Plot Cell Type Distributions</a></li>
  </ul></li>
  <li><a href="#save-the-giotto-objects" id="toc-save-the-giotto-objects" class="nav-link" data-scroll-target="#save-the-giotto-objects"><span class="toc-section-number">10</span>  Save the Giotto Objects</a></li>
  </ul>
</nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Giotto Suite with Xenium Data</h1>
</div>



<div class="quarto-title-meta">

    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-heading">Matthew O'Brien</div>
    <div class="quarto-title-meta-contents">
      <p class="date">August 2, 2023</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<section id="data-and-package-loading" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="data-and-package-loading"><span class="header-section-number">1</span> Data and Package Loading</h2>

In this demo, the <a href="https://www.10xgenomics.com/products/xenium-in-situ/preview-dataset-human-breast">Xenium Pre-Release Human Breast Cancer Dataset</a> from 10X Genomics will be analyzed using Giotto Suite. Only the first tissue replicate is considered in this analysis. First, ensure Giotto Suite is installed and that the data has been unpacked into an appropriate directory. Then, the analysis begins with loading necessary libraries.

<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># ** RESULTS FOLDER **</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>results_folder <span class="ot">=</span> <span class="st">'/projectnb/rd-spat/HOME/mobrien2/xenium/BioC/'</span></span>
<span id="cb1-3"><a href="#cb1-3"></a></span>
<span id="cb1-4"><a href="#cb1-4"></a><span class="fu">library</span>(terra)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>



<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">library</span>(Giotto)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>



<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="fu">library</span>(tictoc)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>terra 1.6.41</code></pre>
<pre><code>Giotto Suite 3.3.1</code></pre>
<pre><code>Attaching package: 'tictoc'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:terra':

    shift, size</code></pre>
</div>
Importanly, <code>GiottoInstructions</code> specify how and where any plotting information will be saved, among other <a href="https://giottosuite.readthedocs.io/en/latest/subsections/getting_started/getting_started_gobject.html#customizing-the-giotto-object">specifications</a>.
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb8-2"><a href="#cb8-2"></a></span>
<span id="cb8-3"><a href="#cb8-3"></a>instrs <span class="ot">=</span> <span class="fu">createGiottoInstructions</span>(<span class="at">save_dir =</span> results_folder,</span>
<span id="cb8-4"><a href="#cb8-4"></a>                                  <span class="at">save_plot =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-5"><a href="#cb8-5"></a>                                  <span class="at">show_plot =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-6"><a href="#cb8-6"></a>                                  <span class="at">return_plot =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>
no external python path was provided, but a giotto python environment was found
 and will be used</code></pre>
</div>

Specify the appropriate path to import the files into R.
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1"></a><span class="co"># ** SET PATH TO FOLDER CONTAINING XENIUM DATA **</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>xenium_folder <span class="ot">=</span> <span class="st">'/projectnb/rd-spat/DATA/Public_data/Spatial/Multiplexing_RNA/xenium/FFPE_Human_Breast_Cancer/Rep1/'</span></span>
<span id="cb10-3"><a href="#cb10-3"></a></span>
<span id="cb10-4"><a href="#cb10-4"></a><span class="co"># General files:</span></span>
<span id="cb10-5"><a href="#cb10-5"></a>settings_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_experiment.xenium'</span>)</span>
<span id="cb10-6"><a href="#cb10-6"></a>he_img_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_he_image.tif'</span>)</span>
<span id="cb10-7"><a href="#cb10-7"></a>if_img_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_if_image.tif'</span>)</span>
<span id="cb10-8"><a href="#cb10-8"></a>panel_meta_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_panel.tsv'</span>) <span class="co"># (optional)</span></span>
<span id="cb10-9"><a href="#cb10-9"></a></span>
<span id="cb10-10"><a href="#cb10-10"></a><span class="co"># Subcellular files:</span></span>
<span id="cb10-11"><a href="#cb10-11"></a>cell_bound_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_cell_boundaries.csv.gz'</span>)</span>
<span id="cb10-12"><a href="#cb10-12"></a>nuc_bound_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_nucleus_boundaries.csv.gz'</span>)</span>
<span id="cb10-13"><a href="#cb10-13"></a>tx_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_transcripts.csv.gz'</span>)</span>
<span id="cb10-14"><a href="#cb10-14"></a>feat_meta_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'cell_feature_matrix/features.tsv.gz'</span>) <span class="co"># (also used in aggregate)</span></span>
<span id="cb10-15"><a href="#cb10-15"></a></span>
<span id="cb10-16"><a href="#cb10-16"></a><span class="co"># Aggregate files:</span></span>
<span id="cb10-17"><a href="#cb10-17"></a>expr_mat_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'cell_feature_matrix'</span>)</span>
<span id="cb10-18"><a href="#cb10-18"></a>cell_meta_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder, <span class="st">'Xenium_FFPE_Human_Breast_Cancer_Rep1_cells.csv.gz'</span>) <span class="co"># contains spatlocs</span></span>
<span id="cb10-19"><a href="#cb10-19"></a></span>
<span id="cb10-20"><a href="#cb10-20"></a><span class="co"># Cell typing files:</span></span>
<span id="cb10-21"><a href="#cb10-21"></a>gene_groups_path <span class="ot">=</span> <span class="fu">paste0</span>(xenium_folder,<span class="st">"Xenium_FFPE_Human_Breast_Cancer_Rep1_gene_groups.csv"</span>)</span>
<span id="cb10-22"><a href="#cb10-22"></a>feat_groups <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">fread</span>(gene_groups_path, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-23"><a href="#cb10-23"></a><span class="fu">colnames</span>(feat_groups) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"feature"</span>, <span class="st">"cell_type"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<details close="">
<summary>Metadata Loading and Exploration of Different Probes for QC</summary>
  
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu"></span></span>
  <span id="cb11-2"><a href="#cb11-2"></a></span>
  <span id="cb11-3"><a href="#cb11-3"></a><span class="co"># Load feature metadata</span></span>
  <span id="cb11-4"><a href="#cb11-4"></a><span class="co"># cell_feature_matrix folder must be unpacked before runtime!</span></span>
  <span id="cb11-5"><a href="#cb11-5"></a></span>
  <span id="cb11-6"><a href="#cb11-6"></a>feature_dt <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">fread</span>(feat_meta_path, <span class="at">header =</span> <span class="cn">FALSE</span>)</span>
  <span id="cb11-7"><a href="#cb11-7"></a><span class="fu">colnames</span>(feature_dt) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'feat_ID'</span>,<span class="st">'feat_name'</span>,<span class="st">'feat_type'</span>)</span>
  <span id="cb11-8"><a href="#cb11-8"></a></span>
  <span id="cb11-9"><a href="#cb11-9"></a><span class="co"># Identify feature IDs that belong to each feature type</span></span>
  <span id="cb11-10"><a href="#cb11-10"></a>feature_dt[, <span class="fu">table</span>(feat_type)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>feat_type
             Blank Codeword           Gene Expression Negative Control Codeword 
                        159                       313                        41 
     Negative Control Probe 
                         28 </code></pre>
  </div>
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>feat_types <span class="ot">=</span> <span class="fu">names</span>(feature_dt[, <span class="fu">table</span>(feat_type)])</span>
  <span id="cb13-2"><a href="#cb13-2"></a></span>
  <span id="cb13-3"><a href="#cb13-3"></a>feat_types_IDs <span class="ot">=</span> <span class="fu">lapply</span>(</span>
  <span id="cb13-4"><a href="#cb13-4"></a>  feat_types, <span class="cf">function</span>(type) feature_dt[feat_type <span class="sc">==</span> type, <span class="fu">unique</span>(feat_ID)]</span>
  <span id="cb13-5"><a href="#cb13-5"></a>)</span>
  <span id="cb13-6"><a href="#cb13-6"></a><span class="fu">names</span>(feat_types_IDs) <span class="ot">=</span> feat_types</span>
  <span id="cb13-7"><a href="#cb13-7"></a></span>
  <span id="cb13-8"><a href="#cb13-8"></a><span class="fu"></span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>

</details>

The path to the transcripts with their respecitve locations will be loaded in using <code>data.table</code>, and some column names are changed for Giotto Suite conventions. The transcripts may then be filtered, here they are filtered according to the Phred score used by 10X. The detected transcripts are then fed into <code>createGiottoPoints()</code>.

<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="fu"></span></span>
<span id="cb15-2"><a href="#cb15-2"></a>tx_dt <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">fread</span>(tx_path)</span>
<span id="cb15-3"><a href="#cb15-3"></a>data.table<span class="sc">::</span><span class="fu">setnames</span>(<span class="at">x =</span> tx_dt,</span>
<span id="cb15-4"><a href="#cb15-4"></a>                     <span class="at">old =</span> <span class="fu">c</span>(<span class="st">'feature_name'</span>, <span class="st">'x_location'</span>, <span class="st">'y_location'</span>),</span>
<span id="cb15-5"><a href="#cb15-5"></a>                     <span class="at">new =</span> <span class="fu">c</span>(<span class="st">'feat_ID'</span>, <span class="st">'x'</span>, <span class="st">'y'</span>))</span>
<span id="cb15-6"><a href="#cb15-6"></a><span class="fu">cat</span>(<span class="st">'Transcripts info available:</span><span class="sc">\n</span><span class="st"> '</span>, <span class="fu">paste0</span>(<span class="st">'"'</span>, <span class="fu">colnames</span>(tx_dt), <span class="st">'"'</span>), <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>,</span>
<span id="cb15-7"><a href="#cb15-7"></a><span class="st">'with'</span>, tx_dt[,.N], <span class="st">'unfiltered detections</span><span class="sc">\n</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Transcripts info available:
  "transcript_id" "cell_id" "overlaps_nucleus" "feat_ID" "x" "y" "z_location" "qv" 
 with 43664530 unfiltered detections</code></pre>
</div>


<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># filter by qv (Phred score)</span></span>
<span id="cb17-2"><a href="#cb17-2"></a>tx_dt_filtered <span class="ot">=</span> tx_dt[qv <span class="sc">&gt;=</span> <span class="dv">20</span>]</span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="fu">cat</span>(<span class="st">'and'</span>, tx_dt_filtered[,.N], <span class="st">'filtered detections</span><span class="sc">\n\n</span><span class="st">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>and 34813341 filtered detections</code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="co"># separate detections by feature type</span></span>
<span id="cb19-2"><a href="#cb19-2"></a>tx_dt_types <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb19-3"><a href="#cb19-3"></a>  feat_types_IDs, <span class="cf">function</span>(types) tx_dt_filtered[feat_ID <span class="sc">%in%</span> types]</span>
<span id="cb19-4"><a href="#cb19-4"></a>)</span>
<span id="cb19-5"><a href="#cb19-5"></a></span>
<span id="cb19-6"><a href="#cb19-6"></a><span class="fu">invisible</span>(<span class="fu">lapply</span>(<span class="fu">seq_along</span>(tx_dt_types), <span class="cf">function</span>(x) {</span>
<span id="cb19-7"><a href="#cb19-7"></a>  <span class="fu">cat</span>(<span class="fu">names</span>(tx_dt_types)[[x]], <span class="st">'detections: '</span>, tx_dt_types[[x]][,.N], <span class="st">'</span><span class="sc">\n</span><span class="st">'</span>)</span>
<span id="cb19-8"><a href="#cb19-8"></a>}))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Blank Codeword detections:  8805 
Gene Expression detections:  34764833 
Negative Control Codeword detections:  1855 
Negative Control Probe detections:  37848 </code></pre>
</div>
</div>

<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu"></span></span>
<span id="cb23-2"><a href="#cb23-2"></a>gpoints_list <span class="ot">=</span> <span class="fu">lapply</span>(</span>
<span id="cb23-3"><a href="#cb23-3"></a>  tx_dt_types, <span class="cf">function</span>(x) <span class="fu">createGiottoPoints</span>(<span class="at">x =</span> x, <span class="at">verbose =</span> F)</span>
<span id="cb23-4"><a href="#cb23-4"></a>) <span class="co"></span></span>
<span id="cb23-5"><a href="#cb23-5"></a></span>
<span id="cb23-6"><a href="#cb23-6"></a><span class="co"># preview QC probe detections</span></span>
<span id="cb23-7"><a href="#cb23-7"></a><span class="fu">plot</span>(gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Blank Codeword</span><span class="st">`</span>,</span>
<span id="cb23-8"><a href="#cb23-8"></a>     <span class="at">point_size =</span> <span class="fl">0.3</span>,</span>
<span id="cb23-9"><a href="#cb23-9"></a>     <span class="at">main =</span> <span class="st">'Blank Codeword'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
<details close="">
<summary>Negative Control Codeword SpatPlot</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="fu">plot</span>(gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Negative Control Codeword</span><span class="st">`</span>,</span>
<span id="cb24-2"><a href="#cb24-2"></a>     <span class="at">point_size =</span> <span class="fl">0.3</span>,</span>
<span id="cb24-3"><a href="#cb24-3"></a>     <span class="at">main =</span> <span class="st">'Negative Control Codeword'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-4-2.png" class="img-fluid" width="672"></p>
</div>
<details close="">
<summary>Negative Control Probe SpatPlot</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="fu">plot</span>(gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Negative Control Probe</span><span class="st">`</span>,</span>
<span id="cb25-2"><a href="#cb25-2"></a>     <span class="at">point_size =</span> <span class="fl">0.3</span>,</span>
<span id="cb25-3"><a href="#cb25-3"></a>     <span class="at">main =</span> <span class="st">'Negative Control Probe'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-4-3.png" class="img-fluid" width="672"></p>
</div>
<details open="">
<summary>KRT8 and MS4A1 Gene Expression</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co"># preview two genes</span></span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="fu">plot</span>(gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Gene Expression</span><span class="st">`</span>,  <span class="co"></span></span>
<span id="cb26-3"><a href="#cb26-3"></a>     <span class="at">feats =</span> <span class="fu">c</span>(<span class="st">'KRT8'</span>, <span class="st">'MS4A1'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-4-4.png" class="img-fluid" width="672"></p>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1"></a>tx_dt_types<span class="sc">$</span><span class="st">`</span><span class="at">Gene Expression</span><span class="st">`</span>[feat_ID <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">'KRT8'</span>, <span class="st">'MS4A1'</span>), <span class="fu">table</span>(feat_ID)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>feat_ID
  KRT8  MS4A1 
530190  20926 </code></pre>
</div>

</div>
</section>
<section id="import-segmentation-data" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="import-segmentation-data"><span class="header-section-number">2</span> Import Segmentation Data</h2>

Now that the transcript data has been imported to the Giotto Obejct in the form of <code>GiottoPoints</code>, the cell segmentation data must be imported in the form of <code>GiottoPolygons</code> so that an expression matrix can be created later on. After importing the <code>boundary.csv.gz</code> files for cell and nucleus segmentations and applying some small naming convention changes, the resulting <code>data.table</code>s can become <code>GiottoPolygons</code>.

<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="fu"></span></span>
<span id="cb31-2"><a href="#cb31-2"></a>cellPoly_dt <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">fread</span>(cell_bound_path)</span>
<span id="cb31-3"><a href="#cb31-3"></a>nucPoly_dt <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">fread</span>(nuc_bound_path)</span>
<span id="cb31-4"><a href="#cb31-4"></a></span>
<span id="cb31-5"><a href="#cb31-5"></a>data.table<span class="sc">::</span><span class="fu">setnames</span>(cellPoly_dt,</span>
<span id="cb31-6"><a href="#cb31-6"></a>                     <span class="at">old =</span> <span class="fu">c</span>(<span class="st">'cell_id'</span>, <span class="st">'vertex_x'</span>, <span class="st">'vertex_y'</span>),</span>
<span id="cb31-7"><a href="#cb31-7"></a>                     <span class="at">new =</span> <span class="fu">c</span>(<span class="st">'poly_ID'</span>, <span class="st">'x'</span>, <span class="st">'y'</span>))</span>
<span id="cb31-8"><a href="#cb31-8"></a>data.table<span class="sc">::</span><span class="fu">setnames</span>(nucPoly_dt,</span>
<span id="cb31-9"><a href="#cb31-9"></a>                     <span class="at">old =</span> <span class="fu">c</span>(<span class="st">'cell_id'</span>, <span class="st">'vertex_x'</span>, <span class="st">'vertex_y'</span>),</span>
<span id="cb31-10"><a href="#cb31-10"></a>                     <span class="at">new =</span> <span class="fu">c</span>(<span class="st">'poly_ID'</span>, <span class="st">'x'</span>, <span class="st">'y'</span>))</span>
<span id="cb31-11"><a href="#cb31-11"></a></span>
<span id="cb31-12"><a href="#cb31-12"></a>gpoly_cells <span class="ot">=</span> <span class="fu">createGiottoPolygonsFromDfr</span>(<span class="at">segmdfr =</span> cellPoly_dt,</span>
<span id="cb31-13"><a href="#cb31-13"></a>                                          <span class="at">name =</span> <span class="st">'cell'</span>,</span>
<span id="cb31-14"><a href="#cb31-14"></a>                                          <span class="at">calc_centroids =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-15"><a href="#cb31-15"></a>                                          <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb31-16"><a href="#cb31-16"></a>gpoly_nucs <span class="ot">=</span> <span class="fu">createGiottoPolygonsFromDfr</span>(<span class="at">segmdfr =</span> nucPoly_dt,</span>
<span id="cb31-17"><a href="#cb31-17"></a>                                         <span class="at">name =</span> <span class="st">'nucleus'</span>,</span>
<span id="cb31-18"><a href="#cb31-18"></a>                                         <span class="at">calc_centroids =</span> <span class="cn">TRUE</span>,</span>
<span id="cb31-19"><a href="#cb31-19"></a>                                         <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb31-20"><a href="#cb31-20"></a><span class="fu"></span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
The <code>GiottoPolygon</code> centroids can be immediately visualized using the plot function.
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a><span class="fu">plot</span>(<span class="at">x =</span> gpoly_nucs, <span class="at">point_size =</span> <span class="fl">0.1</span>, <span class="at">type =</span> <span class="st">'centroid'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="create-a-giotto-object" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="create-a-giotto-object"><span class="header-section-number">3</span> Create a Giotto Object</h2>
The transcript information in the form of <code>GiottoPoints</code> and the polygon information in the form of <code>GiottoPolygons</code> may now be fed into <code>createGiottoObjectSubcellular</code> to create a Giotto Object. 
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">createGiottoObjectSubcellular</span>(</span>
<span id="cb34-2"><a href="#cb34-2"></a>  <span class="at">gpoints =</span> <span class="fu">list</span>(<span class="at">rna =</span> gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Gene Expression</span><span class="st">`</span>,</span>
<span id="cb34-3"><a href="#cb34-3"></a>                 <span class="at">blank_code =</span> gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Blank Codeword</span><span class="st">`</span>,</span>
<span id="cb34-4"><a href="#cb34-4"></a>                 <span class="at">neg_code =</span> gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Negative Control Codeword</span><span class="st">`</span>,</span>
<span id="cb34-5"><a href="#cb34-5"></a>                 <span class="at">neg_probe =</span> gpoints_list<span class="sc">$</span><span class="st">`</span><span class="at">Negative Control Probe</span><span class="st">`</span>),</span>
<span id="cb34-6"><a href="#cb34-6"></a>  <span class="at">gpolygons =</span> <span class="fu">list</span>(<span class="at">cell =</span> gpoly_cells,</span>
<span id="cb34-7"><a href="#cb34-7"></a>                   <span class="at">nucleus =</span> gpoly_nucs),</span>
<span id="cb34-8"><a href="#cb34-8"></a>  <span class="at">instructions =</span> instrs,</span>
<span id="cb34-9"><a href="#cb34-9"></a>  <span class="at">verbose =</span> <span class="cn">FALSE</span></span>
<span id="cb34-10"><a href="#cb34-10"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>polygonlist is a list with names</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[ cell ] Process polygon info...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[ nucleus ] Process polygon info...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>pointslist is a named list</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[ rna ] Process point info...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[ blank_code ] Process point info...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[ neg_code ] Process point info...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>[ neg_probe ] Process point info...</code></pre>
</div>
</div>
<section id="create-a-giotto-object-with-convenience-function" class="level3" data-number="3.1">
<h3 data-number="3.1" class="anchored" data-anchor-id="create-a-giotto-object-with-convenience-function"><span class="header-section-number">3.1</span> Create a Giotto Object with Convenience Function</h3>
Alternatively, Giotto Suite offers a convenience function for reduced data manipulation. 
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">createGiottoXeniumObject</span>(<span class="at">xenium_dir =</span> xenium_folder,</span>
<span id="cb43-2"><a href="#cb43-2"></a>                                       <span class="at">data_to_use =</span> <span class="fu">c</span>(<span class="st">"subcellular"</span>, <span class="st">"aggregate"</span>),</span>
<span id="cb43-3"><a href="#cb43-3"></a>                                       <span class="at">bounds_to_load =</span> <span class="fu">c</span>(<span class="st">"cell"</span>, <span class="st">"nucleus"</span>),</span>
<span id="cb43-4"><a href="#cb43-4"></a>                                       <span class="at">key_list =</span> <span class="fu">list</span>(<span class="at">blank_code =</span> <span class="st">'BLANK_'</span>,</span>
<span id="cb43-5"><a href="#cb43-5"></a>                                                       <span class="at">neg_code =</span> <span class="st">'NegControlCodeword_'</span>,</span>
<span id="cb43-6"><a href="#cb43-6"></a>                                                       <span class="at">neg_probe =</span> <span class="fu">c</span>(<span class="st">'NegControlProbe_|antisense_'</span>)), </span>
<span id="cb43-7"><a href="#cb43-7"></a>                                       <span class="at">instructions =</span> instrs,</span>
<span id="cb43-8"><a href="#cb43-8"></a>                                       <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1"></a><span class="fu">showGiottoSpatialInfo</span>(xenium_gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

<details close="">
<summary>Click to View Spatial Info</summary>
<div class="cell-output cell-output-stdout">
<pre><code>For Spatial info:  cell 

An object of class giottoPolygon
spat_unit : "cell"
Spatial Information:
 class       : SpatVector 
 geometry    : polygons 
 dimensions  : 167782, 1  (geometries, attributes)
 extent      : 0, 7525.9, 0, 5478.038  (xmin, xmax, ymin, ymax)
 coord. ref. :  
 names       : poly_ID
 type        :   &lt;chr&gt;
 values      :       1
                     2
                     3
 centroids   : calculated
 overlaps    : NULL-----------------------------
 
For Spatial info:  nucleus 

An object of class giottoPolygon
spat_unit : "nucleus"
Spatial Information:
 class       : SpatVector 
 geometry    : polygons 
 dimensions  : 167782, 1  (geometries, attributes)
 extent      : 1.4875, 7524.413, 0, 5478.038  (xmin, xmax, ymin, ymax)
 coord. ref. :  
 names       : poly_ID
 type        :   &lt;chr&gt;
 values      :       1
                     2
                     3
 centroids   : calculated
 overlaps    : NULL-----------------------------
 </code></pre>
</div>
</details>
</div>
</section>

</section>
<section id="generate-aggregated-expression-based-on-feature-and-boundary-polygon-information" class="level2" data-number="4">
<h2 data-number="4" class="anchored" data-anchor-id="generate-aggregated-expression-based-on-feature-and-boundary-polygon-information"><span class="header-section-number">4</span> Generate Aggregated Expression based on Feature and Boundary (polygon) Information</h2>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">calculateOverlapRaster</span>(xenium_gobj,</span>
<span id="cb47-2"><a href="#cb47-2"></a>                                     <span class="at">spatial_info =</span> <span class="st">'cell'</span>,</span>
<span id="cb47-3"><a href="#cb47-3"></a>                                     <span class="at">feat_info =</span> <span class="st">'rna'</span>,</span>
<span id="cb47-4"><a href="#cb47-4"></a>                                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb47-5"><a href="#cb47-5"></a></span>
<span id="cb47-6"><a href="#cb47-6"></a>xenium_gobj <span class="ot">=</span> <span class="fu">calculateOverlapRaster</span>(xenium_gobj,</span>
<span id="cb47-7"><a href="#cb47-7"></a>                                     <span class="at">spatial_info =</span> <span class="st">'nucleus'</span>,</span>
<span id="cb47-8"><a href="#cb47-8"></a>                                     <span class="at">feat_info =</span> <span class="st">'rna'</span>,</span>
<span id="cb47-9"><a href="#cb47-9"></a>                                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb47-10"><a href="#cb47-10"></a></span>
<span id="cb47-11"><a href="#cb47-11"></a><span class="fu">showGiottoSpatialInfo</span>(xenium_gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

<details close="">
<summary>Click to View Spatial Info</summary>
<div class="cell-output cell-output-stdout">
<pre><code>For Spatial info:  cell 

An object of class giottoPolygon
spat_unit : "cell"
Spatial Information:
 class       : SpatVector 
 geometry    : polygons 
 dimensions  : 167782, 1  (geometries, attributes)
 extent      : 0, 7525.9, 0, 5478.038  (xmin, xmax, ymin, ymax)
 coord. ref. :  
 names       : poly_ID
 type        :   &lt;chr&gt;
 values      :       1
                     2
                     3
 centroids   : calculated
 overlaps    : calculated-----------------------------
 
For Spatial info:  nucleus 

An object of class giottoPolygon
spat_unit : "nucleus"
Spatial Information:
 class       : SpatVector 
 geometry    : polygons 
 dimensions  : 167782, 1  (geometries, attributes)
 extent      : 1.4875, 7524.413, 0, 5478.038  (xmin, xmax, ymin, ymax)
 coord. ref. :  
 names       : poly_ID
 type        :   &lt;chr&gt;
 values      :       1
                     2
                     3
 centroids   : calculated
 overlaps    : calculated-----------------------------
 </code></pre>
</div>
</details>
</div>
<section id="assign-polygon-overlaps-information-to-expression-matrix" class="level3" data-number="4.1">
<h3 data-number="4.1" class="anchored" data-anchor-id="assign-polygon-overlaps-information-to-expression-matrix"><span class="header-section-number">4.1</span> Assign polygon overlaps information to expression matrix</h3>
<p>In order to create an aggregated expression matrix, the <code>'rna'</code> features overlapped by the <code>'cell'</code> polygon boundaries are sent to be combined into a cell/feature matrix (named as <code>'raw'</code>) in the Giotto object’s <code>expression</code> slot.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">overlapToMatrix</span>(xenium_gobj,</span>
<span id="cb49-2"><a href="#cb49-2"></a>                              <span class="at">poly_info =</span> <span class="st">'cell'</span>,</span>
<span id="cb49-3"><a href="#cb49-3"></a>                              <span class="at">feat_info =</span> <span class="st">'rna'</span>,</span>
<span id="cb49-4"><a href="#cb49-4"></a>                              <span class="at">name =</span> <span class="st">'raw'</span>)</span>
<span id="cb49-5"><a href="#cb49-5"></a></span>
<span id="cb49-6"><a href="#cb49-6"></a>xenium_gobj <span class="ot">=</span> <span class="fu">overlapToMatrix</span>(xenium_gobj,</span>
<span id="cb49-7"><a href="#cb49-7"></a>                              <span class="at">poly_info =</span> <span class="st">'nucleus'</span>,</span>
<span id="cb49-8"><a href="#cb49-8"></a>                              <span class="at">feat_info =</span> <span class="st">'rna'</span>,</span>
<span id="cb49-9"><a href="#cb49-9"></a>                              <span class="at">name =</span> <span class="st">'raw'</span>)</span>
<span id="cb49-10"><a href="#cb49-10"></a></span>
<span id="cb49-11"><a href="#cb49-11"></a><span class="fu">showGiottoExpression</span>(xenium_gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<details close="">
<summary>Click to Preview Calculated Expression</summary>
<div class="cell-output cell-output-stdout">
<pre><code>├──Spatial unit "cell"
│  └──Feature type "rna"
│     └──Expression data "raw" values:
│           An object of class exprObj : "raw"
│           spat_unit : "cell"
│           feat_type : "rna"
│           provenance: cell 
│           
│           contains:
│           313 x 167782 sparse Matrix of class "dgCMatrix"
│                                                  
│           LUM   2 . 3 . 1 1  2 . . 2 . 5 . ......
│           TCIM  1 1 . 4 1 1 13 . . . . . . ......
│           RUNX1 . . . . . .  . . . . 1 . . ......
│           
│            ........suppressing 167769 columns and 307 rows 
│                                                  
│           CD1C   1 . . . . . . . . . . . . ......
│           CYP1A1 . . . . . . . . . . . . . ......
│           CRHBP  . . . . . . . . . . . . . ......
│           
│            First four colnames:
│            1 2 3 4 
│        
└──Spatial unit "nucleus"
   └──Feature type "rna"
      └──Expression data "raw" values:
            An object of class exprObj : "raw"
            spat_unit : "nucleus"
            feat_type : "rna"
            provenance: nucleus 
            
            contains:
            313 x 167782 sparse Matrix of class "dgCMatrix"
                                                  
            LUM   . . . . 1 . 1 . . . . 4 . ......
            TCIM  . . . 3 . 1 9 . . . . . . ......
            RUNX1 . . . . . . . . . . 1 . . ......
            
             ........suppressing 167769 columns and 307 rows 
                                                   
            CD1C   1 . . . . . . . . . . . . ......
            CYP1A1 . . . . . . . . . . . . . ......
            CRHBP  . . . . . . . . . . . . . ......
            
             First four colnames:
             1 2 3 4 
         </code></pre>
</div>
</details>
After creating the expression matrix from the polygon and transcript overlaps, the data may be normalized and clustered in a pipeline similar to that which is depicted below:
<p><img src="xenium_MOB_files/figure-html/norm_dimred_cluster.png" class="img-fluid" width="872"></p>
</div>
</section>

<details close="">
<summary>Adding Metadata, Normalization, and Calculating Highly Variable Features</summary>

  <section id="add-metadata" class="level3" data-number="4.2">
  <h3 data-number="4.2" class="anchored" data-anchor-id="add-metadata"><span class="header-section-number">4.2</span> Add Metadata</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a>panel_meta <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">fread</span>(panel_meta_path)</span>
  <span id="cb51-2"><a href="#cb51-2"></a>data.table<span class="sc">::</span><span class="fu">setnames</span>(panel_meta, <span class="st">'Name'</span>, <span class="st">'feat_ID'</span>)</span>
  <span id="cb51-3"><a href="#cb51-3"></a></span>
  <span id="cb51-4"><a href="#cb51-4"></a><span class="co"># Append this metadata</span></span>
  <span id="cb51-5"><a href="#cb51-5"></a>xenium_gobj <span class="ot">=</span> <span class="fu">addFeatMetadata</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb51-6"><a href="#cb51-6"></a>                              <span class="at">feat_type =</span> <span class="st">'rna'</span>,</span>
  <span id="cb51-7"><a href="#cb51-7"></a>                              <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb51-8"><a href="#cb51-8"></a>                              <span class="at">new_metadata =</span> panel_meta,</span>
  <span id="cb51-9"><a href="#cb51-9"></a>                              <span class="at">by_column =</span> <span class="cn">TRUE</span>,</span>
  <span id="cb51-10"><a href="#cb51-10"></a>                              <span class="at">column_feat_ID =</span> <span class="st">'feat_ID'</span>)</span>
  <span id="cb51-11"><a href="#cb51-11"></a></span>
  <span id="cb51-12"><a href="#cb51-12"></a>xenium_gobj <span class="ot">=</span> <span class="fu">addFeatMetadata</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb51-13"><a href="#cb51-13"></a>                              <span class="at">feat_type =</span> <span class="st">'rna'</span>,</span>
  <span id="cb51-14"><a href="#cb51-14"></a>                              <span class="at">spat_unit =</span> <span class="st">'nucleus'</span>,</span>
  <span id="cb51-15"><a href="#cb51-15"></a>                              <span class="at">new_metadata =</span> panel_meta,</span>
  <span id="cb51-16"><a href="#cb51-16"></a>                              <span class="at">by_column =</span> <span class="cn">TRUE</span>,</span>
  <span id="cb51-17"><a href="#cb51-17"></a>                              <span class="at">column_feat_ID =</span> <span class="st">'feat_ID'</span>)</span>
  <span id="cb51-18"><a href="#cb51-18"></a></span>
  <span id="cb51-19"><a href="#cb51-19"></a><span class="co"># Print all available features metadata</span></span>
  <span id="cb51-20"><a href="#cb51-20"></a><span class="fu">showGiottoFeatMetadata</span>(xenium_gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>├──Spatial unit "cell"
  │  ├──Feature type "rna"
  │  │     An object of class featMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "rna"
  │  │     provenance: cell 
  │  │     
  │  │        feat_ID      Ensembl ID             Annotation
  │  │     1:     LUM ENSG00000139329            Fibroblasts
  │  │     2:    TCIM ENSG00000176907 Breast glandular cells
  │  │     3:   RUNX1 ENSG00000159216          Breast cancer
  │  │  
  │  ├──Feature type "blank_code"
  │  │     An object of class featMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "blank_code"
  │  │     provenance: cell 
  │  │     
  │  │           feat_ID
  │  │     1: BLANK_0424
  │  │     2: BLANK_0401
  │  │     3: BLANK_0447
  │  │  
  │  ├──Feature type "neg_code"
  │  │     An object of class featMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "neg_code"
  │  │     provenance: cell 
  │  │     
  │  │                        feat_ID
  │  │     1: NegControlCodeword_0503
  │  │     2: NegControlCodeword_0514
  │  │     3: NegControlCodeword_0535
  │  │  
  │  └──Feature type "neg_probe"
  │        An object of class featMetaObj 
  │        spat_unit : "cell"
  │        feat_type : "neg_probe"
  │        provenance: cell 
  │        
  │                         feat_ID
  │        1: NegControlProbe_00003
  │        2:       antisense_SCRIB
  │        3: NegControlProbe_00012
  │     
  └──Spatial unit "nucleus"
     ├──Feature type "rna"
     │     An object of class featMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "rna"
     │     provenance: nucleus 
     │     
     │        feat_ID      Ensembl ID             Annotation
     │     1:     LUM ENSG00000139329            Fibroblasts
     │     2:    TCIM ENSG00000176907 Breast glandular cells
     │     3:   RUNX1 ENSG00000159216          Breast cancer
     │  
     ├──Feature type "blank_code"
     │     An object of class featMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "blank_code"
     │     provenance: nucleus 
     │     
     │           feat_ID
     │     1: BLANK_0424
     │     2: BLANK_0401
     │     3: BLANK_0447
     │  
     ├──Feature type "neg_code"
     │     An object of class featMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "neg_code"
     │     provenance: nucleus 
     │     
     │                        feat_ID
     │     1: NegControlCodeword_0503
     │     2: NegControlCodeword_0514
     │     3: NegControlCodeword_0535
     │  
     └──Feature type "neg_probe"
           An object of class featMetaObj 
           spat_unit : "nucleus"
           feat_type : "neg_probe"
           provenance: nucleus 
           
                            feat_ID
           1: NegControlProbe_00003
           2:       antisense_SCRIB
           3: NegControlProbe_00012
        </code></pre>
  </div>
  </div>
  </section>
  <section id="data-filtering" class="level3" data-number="4.3">
  <h3 data-number="4.3" class="anchored" data-anchor-id="data-filtering"><span class="header-section-number">4.3</span> Data filtering</h3>
  <p>Now that an aggregated expression matrix is generated the usual data filtering and processing can be applied. We start by setting a count of 1 to be the minimum to consider a feature expressed. A feature must be detected in at least 3 cells to be included. Lastly, a cell must have a minimum of 5 features detected to be included. <em>Run on a server | <code>229.073 sec elapsed</code></em></p>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">filterGiotto</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb53-2"><a href="#cb53-2"></a>                           <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb53-3"><a href="#cb53-3"></a>                           <span class="at">poly_info =</span> <span class="st">'cell'</span>,</span>
  <span id="cb53-4"><a href="#cb53-4"></a>                           <span class="at">expression_threshold =</span> <span class="dv">1</span>,</span>
  <span id="cb53-5"><a href="#cb53-5"></a>                           <span class="at">feat_det_in_min_cells =</span> <span class="dv">3</span>,</span>
  <span id="cb53-6"><a href="#cb53-6"></a>                           <span class="at">min_det_feats_per_cell =</span> <span class="dv">5</span>,</span>
  <span id="cb53-7"><a href="#cb53-7"></a>                           <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  </div>
  </section>
  <section id="add-data-statistics" class="level3" data-number="4.4">
  <h3 data-number="4.4" class="anchored" data-anchor-id="add-data-statistics"><span class="header-section-number">4.4</span> Add data statistics</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb54"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">addStatistics</span>(xenium_gobj, </span>
  <span id="cb54-2"><a href="#cb54-2"></a>                            <span class="at">expression_values =</span> <span class="st">'raw'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  </div>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb55"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a><span class="fu">showGiottoCellMetadata</span>(xenium_gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>├──Spatial unit "cell"
  │  ├──Feature type "rna"
  │  │     An object of class cellMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "rna"
  │  │     provenance: cell 
  │  │     
  │  │        cell_ID nr_feats perc_feats total_expr
  │  │     1:       1       62   19.80831        156
  │  │     2:       2       41   13.09904         63
  │  │     3:       3       38   12.14058         54
  │  │  
  │  ├──Feature type "blank_code"
  │  │     An object of class cellMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "blank_code"
  │  │     provenance: cell 
  │  │     
  │  │        cell_ID
  │  │     1:       1
  │  │     2:       2
  │  │     3:       3
  │  │  
  │  ├──Feature type "neg_code"
  │  │     An object of class cellMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "neg_code"
  │  │     provenance: cell 
  │  │     
  │  │        cell_ID
  │  │     1:       1
  │  │     2:       2
  │  │     3:       3
  │  │  
  │  └──Feature type "neg_probe"
  │        An object of class cellMetaObj 
  │        spat_unit : "cell"
  │        feat_type : "neg_probe"
  │        provenance: cell 
  │        
  │           cell_ID
  │        1:       1
  │        2:       2
  │        3:       3
  │     
  └──Spatial unit "nucleus"
     ├──Feature type "rna"
     │     An object of class cellMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "rna"
     │     provenance: nucleus 
     │     
     │        cell_ID
     │     1:       1
     │     2:       2
     │     3:       3
     │  
     ├──Feature type "blank_code"
     │     An object of class cellMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "blank_code"
     │     provenance: nucleus 
     │     
     │        cell_ID
     │     1:       1
     │     2:       2
     │     3:       3
     │  
     ├──Feature type "neg_code"
     │     An object of class cellMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "neg_code"
     │     provenance: nucleus 
     │     
     │        cell_ID
     │     1:       1
     │     2:       2
     │     3:       3
     │  
     └──Feature type "neg_probe"
           An object of class cellMetaObj 
           spat_unit : "nucleus"
           feat_type : "neg_probe"
           provenance: nucleus 
           
              cell_ID
           1:       1
           2:       2
           3:       3
        </code></pre>
  </div>
  </div>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="fu">showGiottoFeatMetadata</span>(xenium_gobj)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>├──Spatial unit "cell"
  │  ├──Feature type "rna"
  │  │     An object of class featMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "rna"
  │  │     provenance: cell 
  │  │     
  │  │        feat_ID      Ensembl ID             Annotation nr_cells perc_cells
  │  │     1:     LUM ENSG00000139329            Fibroblasts   101666   61.67669
  │  │     2:    TCIM ENSG00000176907 Breast glandular cells    84842   51.47024
  │  │     3:   RUNX1 ENSG00000159216          Breast cancer    94086   57.07820
  │  │        total_expr mean_expr mean_expr_det
  │  │     1:     946217  5.740319      9.307113
  │  │     2:     300377  1.822267      3.540428
  │  │     3:     229633  1.393091      2.440671
  │  │  
  │  ├──Feature type "blank_code"
  │  │     An object of class featMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "blank_code"
  │  │     provenance: cell 
  │  │     
  │  │           feat_ID
  │  │     1: BLANK_0424
  │  │     2: BLANK_0401
  │  │     3: BLANK_0447
  │  │  
  │  ├──Feature type "neg_code"
  │  │     An object of class featMetaObj 
  │  │     spat_unit : "cell"
  │  │     feat_type : "neg_code"
  │  │     provenance: cell 
  │  │     
  │  │                        feat_ID
  │  │     1: NegControlCodeword_0503
  │  │     2: NegControlCodeword_0514
  │  │     3: NegControlCodeword_0535
  │  │  
  │  └──Feature type "neg_probe"
  │        An object of class featMetaObj 
  │        spat_unit : "cell"
  │        feat_type : "neg_probe"
  │        provenance: cell 
  │        
  │                         feat_ID
  │        1: NegControlProbe_00003
  │        2:       antisense_SCRIB
  │        3: NegControlProbe_00012
  │     
  └──Spatial unit "nucleus"
     ├──Feature type "rna"
     │     An object of class featMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "rna"
     │     provenance: nucleus 
     │     
     │        feat_ID      Ensembl ID             Annotation
     │     1:     LUM ENSG00000139329            Fibroblasts
     │     2:    TCIM ENSG00000176907 Breast glandular cells
     │     3:   RUNX1 ENSG00000159216          Breast cancer
     │  
     ├──Feature type "blank_code"
     │     An object of class featMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "blank_code"
     │     provenance: nucleus 
     │     
     │           feat_ID
     │     1: BLANK_0424
     │     2: BLANK_0401
     │     3: BLANK_0447
     │  
     ├──Feature type "neg_code"
     │     An object of class featMetaObj 
     │     spat_unit : "nucleus"
     │     feat_type : "neg_code"
     │     provenance: nucleus 
     │     
     │                        feat_ID
     │     1: NegControlCodeword_0503
     │     2: NegControlCodeword_0514
     │     3: NegControlCodeword_0535
     │  
     └──Feature type "neg_probe"
           An object of class featMetaObj 
           spat_unit : "nucleus"
           feat_type : "neg_probe"
           provenance: nucleus 
           
                            feat_ID
           1: NegControlProbe_00003
           2:       antisense_SCRIB
           3: NegControlProbe_00012
        </code></pre>
  </div>
  </div>
  </section>
  <section id="normalize-expression" class="level3" data-number="4.5">
  <h3 data-number="4.5" class="anchored" data-anchor-id="normalize-expression"><span class="header-section-number">4.5</span> Normalize expression</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb59"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">normalizeGiotto</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb59-2"><a href="#cb59-2"></a>                              <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb59-3"><a href="#cb59-3"></a>                              <span class="at">scalefactor =</span> <span class="dv">5000</span>,</span>
  <span id="cb59-4"><a href="#cb59-4"></a>                              <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  </div>
  </section>
  <section id="calculate-highly-variable-features" class="level3" data-number="4.6">
  <h3 data-number="4.6" class="anchored" data-anchor-id="calculate-highly-variable-features"><span class="header-section-number">4.6</span> Calculate highly variable features</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb60"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">calculateHVF</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb60-2"><a href="#cb60-2"></a>                           <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb60-3"><a href="#cb60-3"></a>                           <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'2_HVF'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid" width="672"></p>
  </div>
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb61"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a>n_hvf <span class="ot">=</span> <span class="fu">fDataDT</span>(xenium_gobj)[, <span class="fu">sum</span>(hvf <span class="sc">==</span> <span class="st">'yes'</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  </div>
  </section>
  </section>
</details>
<details close="">
<summary>Dimension Reduction and Clustering</summary>
  <section id="dimension-reduction-and-clustering" class="level2" data-number="5">
  <h2 data-number="5" class="anchored" data-anchor-id="dimension-reduction-and-clustering"><span class="header-section-number">5</span> Dimension reduction and clustering</h2>
  <section id="pca" class="level3" data-number="5.1">
  <h3 data-number="5.1" class="anchored" data-anchor-id="pca"><span class="header-section-number">5.1</span> PCA</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb62"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">runPCA</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb62-2"><a href="#cb62-2"></a>                     <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb62-3"><a href="#cb62-3"></a>                     <span class="at">expression_values =</span> <span class="st">'scaled'</span>,</span>
  <span id="cb62-4"><a href="#cb62-4"></a>                     <span class="at">feats_to_use =</span> <span class="cn">NULL</span>,</span>
  <span id="cb62-5"><a href="#cb62-5"></a>                     <span class="at">scale_unit =</span> F,</span>
  <span id="cb62-6"><a href="#cb62-6"></a>                     <span class="at">center =</span> F,</span>
  <span id="cb62-7"><a href="#cb62-7"></a>                     <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>[1] "finished runPCA_factominer, method == factominer"</code></pre>
  </div>
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb64"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a><span class="co"># Visualize Screeplot and PCA</span></span>
  <span id="cb64-2"><a href="#cb64-2"></a><span class="fu">screePlot</span>(xenium_gobj,</span>
  <span id="cb64-3"><a href="#cb64-3"></a>          <span class="at">ncp =</span> <span class="dv">20</span>,</span>
  <span id="cb64-4"><a href="#cb64-4"></a>          <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'3a_screePlot'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>PCA with name:  pca  already exists and will be used for the screeplot </code></pre>
  </div>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
  </div>
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb66"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a><span class="fu">plotPCA</span>(xenium_gobj,</span>
  <span id="cb66-2"><a href="#cb66-2"></a>        <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb66-3"><a href="#cb66-3"></a>        <span class="at">dim_reduction_name =</span> <span class="st">'pca'</span>,</span>
  <span id="cb66-4"><a href="#cb66-4"></a>        <span class="at">point_size =</span> <span class="fl">0.01</span>,</span>
  <span id="cb66-5"><a href="#cb66-5"></a>        <span class="at">dim1_to_use =</span> <span class="dv">1</span>,</span>
  <span id="cb66-6"><a href="#cb66-6"></a>        <span class="at">dim2_to_use =</span> <span class="dv">2</span>,</span>
  <span id="cb66-7"><a href="#cb66-7"></a>        <span class="at">title =</span> <span class="st">'PCA'</span>,</span>
  <span id="cb66-8"><a href="#cb66-8"></a>        <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'3b_PCA'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-19-2.png" class="img-fluid" width="672"></p>
  </div>
  </div>
  </section>
  <section id="tsne-and-umap" class="level3" data-number="5.2">
  <h3 data-number="5.2" class="anchored" data-anchor-id="tsne-and-umap"><span class="header-section-number">5.2</span> tSNE and UMAP</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">runtSNE</span>(xenium_gobj,</span>
  <span id="cb67-2"><a href="#cb67-2"></a>                      <span class="at">spat_unit =</span> <span class="st">'cell'</span>, </span>
  <span id="cb67-3"><a href="#cb67-3"></a>                      <span class="at">dimensions_to_use =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,</span>
  <span id="cb67-4"><a href="#cb67-4"></a>                      <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
  <span id="cb67-5"><a href="#cb67-5"></a></span>
  <span id="cb67-6"><a href="#cb67-6"></a><span class="fu">plotTSNE</span>(xenium_gobj,</span>
  <span id="cb67-7"><a href="#cb67-7"></a>         <span class="at">point_size =</span> <span class="fl">0.01</span>,</span>
  <span id="cb67-8"><a href="#cb67-8"></a>         <span class="at">title =</span> <span class="st">'tSNE'</span>,</span>
  <span id="cb67-9"><a href="#cb67-9"></a>         <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'4a_tSNE'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-20-1.png" class="img-fluid" width="672"></p>
  </div>
  </div>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb68"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">runUMAP</span>(xenium_gobj,</span>
  <span id="cb68-2"><a href="#cb68-2"></a>                      <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb68-3"><a href="#cb68-3"></a>                      <span class="at">dimensions_to_use =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,</span>
  <span id="cb68-4"><a href="#cb68-4"></a>                      <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
  <span id="cb68-5"><a href="#cb68-5"></a></span>
  <span id="cb68-6"><a href="#cb68-6"></a><span class="co"># UMAP in three dimensions</span></span>
  <span id="cb68-7"><a href="#cb68-7"></a>xenium_gobj <span class="ot">=</span> <span class="fu">runUMAP</span>(xenium_gobj,</span>
  <span id="cb68-8"><a href="#cb68-8"></a>                      <span class="at">spat_unit =</span> <span class="st">'cell'</span>, </span>
  <span id="cb68-9"><a href="#cb68-9"></a>                      <span class="at">dimensions_to_use =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,</span>
  <span id="cb68-10"><a href="#cb68-10"></a>                      <span class="at">n_components =</span> <span class="dv">3</span>,</span>
  <span id="cb68-11"><a href="#cb68-11"></a>                      <span class="at">name =</span> <span class="st">'umap_3D'</span>,</span>
  <span id="cb68-12"><a href="#cb68-12"></a>                      <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
  <span id="cb68-13"><a href="#cb68-13"></a></span>
  <span id="cb68-14"><a href="#cb68-14"></a><span class="fu">plotUMAP</span>(xenium_gobj,</span>
  <span id="cb68-15"><a href="#cb68-15"></a>         <span class="at">point_size =</span> <span class="fl">0.01</span>,</span>
  <span id="cb68-16"><a href="#cb68-16"></a>         <span class="at">title =</span> <span class="st">'UMAP'</span>,</span>
  <span id="cb68-17"><a href="#cb68-17"></a>         <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'4b_UMAP'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid" width="672"></p>
  </div>
  </div>
  </section>
  <section id="snn-and-leiden-clustering" class="level3" data-number="5.3">
  <h3 data-number="5.3" class="anchored" data-anchor-id="snn-and-leiden-clustering"><span class="header-section-number">5.3</span> sNN and Leiden clustering</h3>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb69"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1"></a>xenium_gobj <span class="ot">=</span> <span class="fu">createNearestNetwork</span>(xenium_gobj,</span>
  <span id="cb69-2"><a href="#cb69-2"></a>                                   <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb69-3"><a href="#cb69-3"></a>                                   <span class="at">feat_type =</span> <span class="st">'rna'</span>,</span>
  <span id="cb69-4"><a href="#cb69-4"></a>                                   <span class="at">dimensions_to_use =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>,</span>
  <span id="cb69-5"><a href="#cb69-5"></a>                                   <span class="at">k =</span> <span class="dv">10</span>,</span>
  <span id="cb69-6"><a href="#cb69-6"></a>                                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
  <span id="cb69-7"><a href="#cb69-7"></a></span>
  <span id="cb69-8"><a href="#cb69-8"></a>xenium_gobj <span class="ot">=</span> <span class="fu">doLeidenCluster</span>(xenium_gobj,</span>
  <span id="cb69-9"><a href="#cb69-9"></a>                              <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb69-10"><a href="#cb69-10"></a>                              <span class="at">feat_type =</span> <span class="st">'rna'</span>,</span>
  <span id="cb69-11"><a href="#cb69-11"></a>                              <span class="at">resolution =</span> <span class="fl">0.25</span>,</span>
  <span id="cb69-12"><a href="#cb69-12"></a>                              <span class="at">n_iterations =</span> <span class="dv">100</span>)</span>
  <span id="cb69-13"><a href="#cb69-13"></a></span>
  <span id="cb69-14"><a href="#cb69-14"></a><span class="co"># visualize UMAP cluster results</span></span>
  <span id="cb69-15"><a href="#cb69-15"></a><span class="fu">plotUMAP</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb69-16"><a href="#cb69-16"></a>         <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb69-17"><a href="#cb69-17"></a>         <span class="at">cell_color =</span> <span class="st">'leiden_clus'</span>,</span>
  <span id="cb69-18"><a href="#cb69-18"></a>         <span class="at">show_legend =</span> <span class="cn">FALSE</span>,</span>
  <span id="cb69-19"><a href="#cb69-19"></a>         <span class="at">point_size =</span> <span class="fl">0.01</span>,</span>
  <span id="cb69-20"><a href="#cb69-20"></a>         <span class="at">point_shape =</span> <span class="st">'no_border'</span>,</span>
  <span id="cb69-21"><a href="#cb69-21"></a>         <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'5a_umap_leiden'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-22-1.png" class="img-fluid" width="672"></p>
  </div>
  </div>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1"></a><span class="fu">plotUMAP_3D</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb70-2"><a href="#cb70-2"></a>         <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb70-3"><a href="#cb70-3"></a>         <span class="at">cell_color =</span> <span class="st">'leiden_clus'</span>,</span>
  <span id="cb70-4"><a href="#cb70-4"></a>         <span class="at">dim_reduction_name =</span> <span class="st">'umap_3D'</span>,</span>
  <span id="cb70-5"><a href="#cb70-5"></a>         <span class="at">point_size =</span> <span class="fl">0.01</span>,</span>
  <span id="cb70-6"><a href="#cb70-6"></a>         <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'5b_umap_leiden_3D'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stdout">
  <pre><code>create 3D plot
   center label is not clear to see in 3D plot
   You can shut it down with show_center_label = F</code></pre>
  </div>
  </div>

  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb72"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1"></a><span class="fu">spatPlot2D</span>(<span class="at">gobject =</span> xenium_gobj,</span>
  <span id="cb72-2"><a href="#cb72-2"></a>           <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
  <span id="cb72-3"><a href="#cb72-3"></a>           <span class="at">cell_color =</span> <span class="st">'leiden_clus'</span>,</span>
  <span id="cb72-4"><a href="#cb72-4"></a>           <span class="at">point_size =</span> <span class="fl">0.1</span>,</span>
  <span id="cb72-5"><a href="#cb72-5"></a>           <span class="at">point_shape =</span> <span class="st">'no_border'</span>,</span>
  <span id="cb72-6"><a href="#cb72-6"></a>           <span class="at">background_color =</span> <span class="st">'black'</span>,</span>
  <span id="cb72-7"><a href="#cb72-7"></a>           <span class="at">show_legend =</span> <span class="cn">TRUE</span>,</span>
  <span id="cb72-8"><a href="#cb72-8"></a>           <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'6_spat_leiden'</span>,</span>
  <span id="cb72-9"><a href="#cb72-9"></a>                             <span class="at">base_width =</span> <span class="dv">15</span>,</span>
  <span id="cb72-10"><a href="#cb72-10"></a>                             <span class="at">base_height =</span> <span class="dv">15</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid" width="672"></p>
  </div>
  </div>
  <div class="cell">
  <details open="">
  <summary>Code</summary>
  <div class="sourceCode cell-code" id="cb73"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1"></a><span class="fu">spatInSituPlotPoints</span>(xenium_gobj,</span>
  <span id="cb73-2"><a href="#cb73-2"></a>                     <span class="at">show_image =</span> <span class="cn">FALSE</span>,</span>
  <span id="cb73-3"><a href="#cb73-3"></a>                     <span class="at">feats =</span> <span class="cn">NULL</span>,</span>
  <span id="cb73-4"><a href="#cb73-4"></a>                     <span class="at">point_size =</span> <span class="fl">0.05</span>,</span>
  <span id="cb73-5"><a href="#cb73-5"></a>                     <span class="at">show_polygon =</span> <span class="cn">TRUE</span>,</span>
  <span id="cb73-6"><a href="#cb73-6"></a>                     <span class="at">polygon_feat_type =</span> <span class="st">'cell'</span>,</span>
  <span id="cb73-7"><a href="#cb73-7"></a>                     <span class="at">polygon_alpha =</span> <span class="dv">1</span>,</span>
  <span id="cb73-8"><a href="#cb73-8"></a>                     <span class="at">polygon_color =</span> <span class="st">'black'</span>,</span>
  <span id="cb73-9"><a href="#cb73-9"></a>                     <span class="at">polygon_line_size =</span> <span class="fl">0.01</span>,</span>
  <span id="cb73-10"><a href="#cb73-10"></a>                     <span class="at">polygon_fill =</span> <span class="st">'leiden_clus'</span>,</span>
  <span id="cb73-11"><a href="#cb73-11"></a>                     <span class="at">polygon_fill_as_factor =</span> <span class="cn">TRUE</span>,</span>
  <span id="cb73-12"><a href="#cb73-12"></a>                     <span class="at">coord_fix_ratio =</span> <span class="cn">TRUE</span>,</span>
  <span id="cb73-13"><a href="#cb73-13"></a>                     <span class="at">save_para =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'7_polys'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
  </details>
  <div class="cell-output cell-output-stderr">
  <pre><code>Warning in spatInSituPlotPoints(xenium_gobj, show_image = FALSE, feats = NULL, : You need to select features (feats) and modify feature types (feat_type) if you want to show individual features (e.g. transcripts) </code></pre>
  </div>
  <div class="cell-output cell-output-stderr">
  <pre><code>plot polygon layer done</code></pre>
  </div>
  <div class="cell-output-display">
  <p><img src="xenium_MOB_files/figure-html/unnamed-chunk-25-1.png" class="img-fluid" width="672"></p>
  </div>
  </div>
  </section>
  </section>
</details>
<section id="spatial-expression-patterns" class="level2" data-number="6">
<h2 data-number="6" class="anchored" data-anchor-id="spatial-expression-patterns"><span class="header-section-number">6</span> Spatial expression patterns</h2>
<p>Cell centroid distances are utilized to create spatial networks using Delaunay Triangulation. Minimum neighbors and maximum distances may be specified, and the resulting network may be visualized spatially.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1"></a><span class="co"># Generate a Spatial Delaunay Network between cell centroids</span></span>
<span id="cb76-2"><a href="#cb76-2"></a></span>
<span id="cb76-3"><a href="#cb76-3"></a>xenium_gobj <span class="ot">=</span> <span class="fu">createSpatialNetwork</span>(xenium_gobj,</span>
<span id="cb76-4"><a href="#cb76-4"></a>                                   <span class="at">spat_unit =</span> <span class="st">"cell"</span>,</span>
<span id="cb76-5"><a href="#cb76-5"></a>                                   <span class="at">feat_type =</span> <span class="st">"rna"</span>,</span>
<span id="cb76-6"><a href="#cb76-6"></a>                                   <span class="at">minimum_k =</span> <span class="dv">2</span>,</span>
<span id="cb76-7"><a href="#cb76-7"></a>                                   <span class="at">maximum_distance_delaunay =</span> <span class="dv">50</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1"></a><span class="fu">spatPlot2D</span>(xenium_gobj,</span>
<span id="cb77-2"><a href="#cb77-2"></a>           <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
<span id="cb77-3"><a href="#cb77-3"></a>           <span class="at">point_shape =</span> <span class="st">'no_border'</span>,</span>
<span id="cb77-4"><a href="#cb77-4"></a>           <span class="at">show_network =</span> <span class="cn">TRUE</span>, </span>
<span id="cb77-5"><a href="#cb77-5"></a>           <span class="at">point_size =</span> <span class="fl">0.01</span>,</span>
<span id="cb77-6"><a href="#cb77-6"></a>           <span class="at">point_alpha =</span> <span class="fl">0.4</span>,</span>
<span id="cb77-7"><a href="#cb77-7"></a>           <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">base_width =</span> <span class="dv">7</span>,</span>
<span id="cb77-8"><a href="#cb77-8"></a>                             <span class="at">base_height =</span> <span class="dv">7</span>,</span>
<span id="cb77-9"><a href="#cb77-9"></a>                             <span class="at">save_name =</span> <span class="st">'8_spatial_network'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-27-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="view-images" class="level2" data-number="7">
<h2 data-number="7" class="anchored" data-anchor-id="view-images"><span class="header-section-number">7</span> View images</h2>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1"></a>he_img <span class="ot">=</span> <span class="fu">createGiottoLargeImage</span>(<span class="at">raster_object =</span> he_img_path,</span>
<span id="cb78-2"><a href="#cb78-2"></a>                                <span class="at">name =</span> <span class="st">'he'</span>)</span>
<span id="cb78-3"><a href="#cb78-3"></a></span>
<span id="cb78-4"><a href="#cb78-4"></a><span class="fu">plot</span>(he_img)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-28-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1"></a>if_img <span class="ot">=</span> <span class="fu">createGiottoLargeImage</span>(<span class="at">raster_object =</span> if_img_path,</span>
<span id="cb79-2"><a href="#cb79-2"></a>                                <span class="at">name =</span> <span class="st">'if'</span>)</span>
<span id="cb79-3"><a href="#cb79-3"></a></span>
<span id="cb79-4"><a href="#cb79-4"></a><span class="fu">plot</span>(if_img)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-29-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="subsetting-the-giotto-object" class="level2" data-number="8">
<h2 data-number="8" class="anchored" data-anchor-id="subsetting-the-giotto-object"><span class="header-section-number">8</span> Subsetting the Giotto Object</h2>

Subsetting subcellular data is especially useful for enhanced visualizations, as well as faster computations. Here, we will subset to an arbitrary ROI and visualize the polygons and transcripts from the original data provided by 10X Genomics.

<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1"></a>xen_subset <span class="ot">=</span> <span class="fu">subsetGiottoLocs</span>(xenium_gobj,</span>
<span id="cb80-2"><a href="#cb80-2"></a>                              <span class="at">x_max =</span> <span class="dv">7000</span>,</span>
<span id="cb80-3"><a href="#cb80-3"></a>                              <span class="at">x_min =</span> <span class="dv">6100</span>,</span>
<span id="cb80-4"><a href="#cb80-4"></a>                              <span class="at">y_max =</span> <span class="dv">5000</span>,</span>
<span id="cb80-5"><a href="#cb80-5"></a>                              <span class="at">y_min =</span> <span class="dv">3600</span>,</span>
<span id="cb80-6"><a href="#cb80-6"></a>                              <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb80-7"><a href="#cb80-7"></a></span>
<span id="cb80-8"><a href="#cb80-8"></a><span class="fu">spatInSituPlotPoints</span>(xen_subset,</span>
<span id="cb80-9"><a href="#cb80-9"></a>                     <span class="at">show_image =</span> <span class="cn">FALSE</span>,</span>
<span id="cb80-10"><a href="#cb80-10"></a>                     <span class="at">feats =</span> <span class="fu">list</span>(<span class="st">'rna'</span> <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"LUM"</span>, <span class="st">"CXCR4"</span>, <span class="st">"ITGAX"</span>)),</span>
<span id="cb80-11"><a href="#cb80-11"></a>                     <span class="at">feats_color_code =</span> <span class="fu">c</span>(<span class="st">"LUM"</span> <span class="ot">=</span> <span class="st">'green'</span>, <span class="st">'CXCR4'</span> <span class="ot">=</span> <span class="st">'blue'</span>, <span class="st">'ITGAX'</span> <span class="ot">=</span> <span class="st">'red'</span>),</span>
<span id="cb80-12"><a href="#cb80-12"></a>                     <span class="at">point_size =</span> <span class="fl">0.05</span>,</span>
<span id="cb80-13"><a href="#cb80-13"></a>                     <span class="at">show_polygon =</span> <span class="cn">TRUE</span>,</span>
<span id="cb80-14"><a href="#cb80-14"></a>                     <span class="at">polygon_feat_type =</span> <span class="st">'cell'</span>,</span>
<span id="cb80-15"><a href="#cb80-15"></a>                     <span class="at">polygon_color =</span> <span class="st">'black'</span>,</span>
<span id="cb80-16"><a href="#cb80-16"></a>                     <span class="at">polygon_line_size =</span> <span class="fl">0.01</span>,</span>
<span id="cb80-17"><a href="#cb80-17"></a>                     <span class="at">polygon_fill =</span> <span class="st">'leiden_clus'</span>,</span>
<span id="cb80-18"><a href="#cb80-18"></a>                     <span class="at">polygon_fill_as_factor =</span> <span class="cn">TRUE</span>,</span>
<span id="cb80-19"><a href="#cb80-19"></a>                     <span class="at">coord_fix_ratio =</span> <span class="cn">TRUE</span>,</span>
<span id="cb80-20"><a href="#cb80-20"></a>                     <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">save_name =</span> <span class="st">'9_subset_in_situ'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>--| Plotting 40644 feature points</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>plot feature points layer done</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>plot polygon layer done</code></pre>
</div>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid" width="672"></p>
</div>
</div>

Subsetting preserves the structures created previously, such as the spatial network:

<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">spatPlot2D</span>(xen_subset,</span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">spat_unit =</span> <span class="st">'cell'</span>,</span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">point_shape =</span> <span class="st">'no_border'</span>,</span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">show_network =</span> <span class="cn">TRUE</span>, </span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">point_size =</span> <span class="fl">0.5</span>,</span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>            <span class="at">point_alpha =</span> <span class="fl">0.75</span>,</span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>            <span class="at">save_param =</span> <span class="fu">list</span>(<span class="at">base_width =</span> <span class="dv">7</span>,</span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>                              <span class="at">base_height =</span> <span class="dv">7</span>,</span>
<span id="cb80-9"><a href="#cb80-9" aria-hidden="true" tabindex="-1"></a>                              <span class="at">save_name =</span> <span class="st">'8_spatial_network'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/sub_spat_net.png" class="img-fluid" width="672"></p>
</div>

</section>
<section id="custom-segmentations" class="level2" data-number="9">
<h2 data-number="9" class="anchored" data-anchor-id="custom-segmentations"><span class="header-section-number">9</span> Custom Segmentation(s)</h2>

Alternative segmentation methods may also be imported to the Giotto Object, if desired. Here, polygon annotations from <code>StarDist</code>, a <code>QuPath</code> extension, are imported through <code>terra</code>. They will function with the same pipeline as the original data!

<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1"></a><span class="co"># Import</span></span>
<span id="cb84-2"><a href="#cb84-2"></a>star_dist <span class="ot">=</span> terra<span class="sc">::</span><span class="fu">vect</span>(<span class="st">"/projectnb/rd-spat/HOME/mobrien2/xenium/segmentations/imageJ_if_image.geojson"</span>)</span>
<span id="cb84-3"><a href="#cb84-3"></a></span>
<span id="cb84-4"><a href="#cb84-4"></a><span class="co"># Flip orientation</span></span>
<span id="cb84-5"><a href="#cb84-5"></a>star_dist <span class="ot">=</span> terra<span class="sc">::</span><span class="fu">flip</span>(star_dist, <span class="at">direction =</span> <span class="st">"vertical"</span>)</span>
<span id="cb84-6"><a href="#cb84-6"></a>star_dist <span class="ot">=</span> terra<span class="sc">::</span><span class="fu">flip</span>(star_dist, <span class="at">direction =</span> <span class="st">"horizontal"</span>)</span>
<span id="cb84-7"><a href="#cb84-7"></a></span>
<span id="cb84-8"><a href="#cb84-8"></a><span class="co"># Manual manipulation of scaling </span></span>
<span id="cb84-9"><a href="#cb84-9"></a>orig_ext <span class="ot">=</span> <span class="fu">ext</span>(star_dist)<span class="sc">@</span>ptr<span class="sc">$</span>vector</span>
<span id="cb84-10"><a href="#cb84-10"></a>orig_xrange <span class="ot">=</span> orig_ext[[<span class="dv">2</span>]] <span class="sc">-</span> orig_ext[[<span class="dv">1</span>]]</span>
<span id="cb84-11"><a href="#cb84-11"></a>orig_yrange <span class="ot">=</span> orig_ext[[<span class="dv">4</span>]] <span class="sc">-</span> orig_ext[[<span class="dv">3</span>]]</span>
<span id="cb84-12"><a href="#cb84-12"></a></span>
<span id="cb84-13"><a href="#cb84-13"></a>target_xrange <span class="ot">=</span> <span class="dv">6945</span> <span class="sc">-</span> <span class="dv">6115</span></span>
<span id="cb84-14"><a href="#cb84-14"></a>target_yrange <span class="ot">=</span> <span class="dv">4885</span> <span class="sc">-</span> <span class="dv">3915</span></span>
<span id="cb84-15"><a href="#cb84-15"></a></span>
<span id="cb84-16"><a href="#cb84-16"></a>x_adj <span class="ot">=</span> target_xrange<span class="sc">/</span>orig_xrange</span>
<span id="cb84-17"><a href="#cb84-17"></a>y_adj <span class="ot">=</span> target_yrange<span class="sc">/</span>orig_yrange</span>
<span id="cb84-18"><a href="#cb84-18"></a></span>
<span id="cb84-19"><a href="#cb84-19"></a>star_dist <span class="ot">=</span> terra<span class="sc">::</span><span class="fu">rescale</span>(star_dist, <span class="at">fx =</span> x_adj, <span class="at">fy =</span> y_adj)</span>
<span id="cb84-20"><a href="#cb84-20"></a></span>
<span id="cb84-21"><a href="#cb84-21"></a>new_ext <span class="ot">=</span> <span class="fu">ext</span>(star_dist)<span class="sc">@</span>ptr<span class="sc">$</span>vector</span>
<span id="cb84-22"><a href="#cb84-22"></a></span>
<span id="cb84-23"><a href="#cb84-23"></a>x_shift_adj <span class="ot">=</span> <span class="dv">6115</span> <span class="sc">-</span> new_ext[[<span class="dv">1</span>]]</span>
<span id="cb84-24"><a href="#cb84-24"></a>y_shift_adj <span class="ot">=</span> <span class="dv">3915</span> <span class="sc">-</span> new_ext[[<span class="dv">3</span>]]</span>
<span id="cb84-25"><a href="#cb84-25"></a></span>
<span id="cb84-26"><a href="#cb84-26"></a></span>
<span id="cb84-27"><a href="#cb84-27"></a><span class="co"># Manipulate extent</span></span>
<span id="cb84-28"><a href="#cb84-28"></a>star_dist <span class="ot">=</span> terra<span class="sc">::</span><span class="fu">shift</span>(star_dist, <span class="at">dx =</span> x_shift_adj, <span class="at">dy =</span> y_shift_adj)</span>
<span id="cb84-29"><a href="#cb84-29"></a></span>
<span id="cb84-30"><a href="#cb84-30"></a></span>
<span id="cb84-31"><a href="#cb84-31"></a><span class="fu">names</span>(star_dist)[[<span class="dv">1</span>]] <span class="ot">=</span> <span class="st">"poly_ID"</span></span>
<span id="cb84-32"><a href="#cb84-32"></a></span>
<span id="cb84-33"><a href="#cb84-33"></a><span class="co"># Add to Giotto Object</span></span>
<span id="cb84-34"><a href="#cb84-34"></a>sd_polyg <span class="ot">=</span> <span class="fu">giottoPolygon</span>(<span class="at">spatVector =</span> star_dist, </span>
<span id="cb84-35"><a href="#cb84-35"></a>                         <span class="at">unique_ID_cache =</span> <span class="fu">names</span>(star_dist)[[<span class="dv">1</span>]], </span>
<span id="cb84-36"><a href="#cb84-36"></a>                         <span class="at">name =</span> <span class="st">"StarDist"</span>)</span>
<span id="cb84-37"><a href="#cb84-37"></a></span>
<span id="cb84-38"><a href="#cb84-38"></a>xen_subset <span class="ot">=</span> <span class="fu">addGiottoPolygons</span>(xen_subset, </span>
<span id="cb84-39"><a href="#cb84-39"></a>                               <span class="at">gpolygons =</span> <span class="fu">list</span>(sd_polyg))</span>
<span id="cb84-40"><a href="#cb84-40"></a></span>
<span id="cb84-41"><a href="#cb84-41"></a>xen_subset <span class="ot">=</span> <span class="fu">addSpatialCentroidLocations</span>(xen_subset, </span>
<span id="cb84-42"><a href="#cb84-42"></a>                                         <span class="at">poly_info =</span> <span class="st">"StarDist"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Start centroid calculation for polygon information layer: StarDist</code></pre>
</div>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1"></a><span class="co"># Calculate Expression from new set of Polygons</span></span>
<span id="cb86-2"><a href="#cb86-2"></a></span>
<span id="cb86-3"><a href="#cb86-3"></a>xen_subset <span class="ot">=</span> <span class="fu">calculateOverlapRaster</span>(xen_subset, </span>
<span id="cb86-4"><a href="#cb86-4"></a>                             <span class="at">spatial_info =</span> <span class="st">"StarDist"</span>, </span>
<span id="cb86-5"><a href="#cb86-5"></a>                             <span class="at">feat_info =</span> <span class="st">"rna"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>1. convert polygon to raster 
2. overlap raster and points 
3. add polygon information 
4. add points information 
5. create overlap polygon information </code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1"></a>xen_subset <span class="ot">=</span> <span class="fu">overlapToMatrix</span>(xen_subset,</span>
<span id="cb88-2"><a href="#cb88-2"></a>                      <span class="at">name =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb88-3"><a href="#cb88-3"></a>                      <span class="at">poly_info =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb88-4"><a href="#cb88-4"></a>                      <span class="at">feat_info =</span> <span class="st">"rna"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details close="">
<summary>Filter and Normalize</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1"></a>xen_subset <span class="ot">=</span> <span class="fu">filterGiotto</span>(xen_subset,</span>
<span id="cb89-2"><a href="#cb89-2"></a>                   <span class="at">spat_unit =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb89-3"><a href="#cb89-3"></a>                   <span class="at">feat_type =</span> <span class="st">"rna"</span>,</span>
<span id="cb89-4"><a href="#cb89-4"></a>                   <span class="at">expression_values =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb89-5"><a href="#cb89-5"></a>                   <span class="at">poly_info =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb89-6"><a href="#cb89-6"></a>                   <span class="at">expression_threshold =</span> <span class="dv">1</span>,</span>
<span id="cb89-7"><a href="#cb89-7"></a>                   <span class="at">feat_det_in_min_cells =</span> <span class="dv">3</span>,</span>
<span id="cb89-8"><a href="#cb89-8"></a>                   <span class="at">min_det_feats_per_cell =</span> <span class="dv">5</span>,</span>
<span id="cb89-9"><a href="#cb89-9"></a>                   <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb89-10"><a href="#cb89-10"></a></span>
<span id="cb89-11"><a href="#cb89-11"></a>xen_subset <span class="ot">=</span> <span class="fu">normalizeGiotto</span>(xen_subset, </span>
<span id="cb89-12"><a href="#cb89-12"></a>                      <span class="at">expression_values =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb89-13"><a href="#cb89-13"></a>                      <span class="at">spat_unit =</span> <span class="st">"StarDist"</span>, </span>
<span id="cb89-14"><a href="#cb89-14"></a>                      <span class="at">feat_type =</span> <span class="st">"rna"</span>,</span>
<span id="cb89-15"><a href="#cb89-15"></a>                      <span class="at">norm_methods =</span> <span class="st">"standard"</span>,</span>
<span id="cb89-16"><a href="#cb89-16"></a>                      <span class="at">scalefactor =</span> <span class="dv">5000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

<details close="">
<summary>Add Statistics</summary>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1"></a>xen_subset <span class="ot">=</span> <span class="fu">addStatistics</span>(xen_subset, </span>
<span id="cb91-2"><a href="#cb91-2"></a>                    <span class="at">spat_unit =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb91-3"><a href="#cb91-3"></a>                    <span class="at">feat_type =</span> <span class="st">"rna"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<section id="cell-typing-for-each-segmentation-method-by-page-enrichment" class="level3" data-number="9.1">
<h3 data-number="9.1" class="anchored" data-anchor-id="cell-typing-for-each-segmentation-method-by-page-enrichment"><span class="header-section-number">9.1</span> Cell Typing for each Segmentation Method by PAGE Enrichment</h3>
The provided genes and associated cell types provided by 10X Genomics for this data can be used to identify the cell types in this subset. To do so, a cell-type enrichment can be run on each set of segmentations. A Parametric Analysis of Gene Set Enrichment (PAGE) is run here. Giotto Suite's PAGE implementation calculates either Z-scores or -log10(p-values) for a gene set based on fold change. A sign matrix must be provided to the function; it is a binarized matrix of genes (rows) by cell types (columns) in which a value of 1 indicates association between given gene and cell type.
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1"></a>provided_cell_types <span class="ot">=</span> <span class="fu">unique</span>(feat_groups[, cell_type])</span>
<span id="cb92-2"><a href="#cb92-2"></a>sign_list <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb92-3"><a href="#cb92-3"></a>idx <span class="ot">=</span> <span class="dv">0</span></span>
<span id="cb92-4"><a href="#cb92-4"></a><span class="cf">for</span> (ct <span class="cf">in</span> provided_cell_types){</span>
<span id="cb92-5"><a href="#cb92-5"></a>  idx <span class="ot">=</span> idx <span class="sc">+</span> <span class="dv">1</span></span>
<span id="cb92-6"><a href="#cb92-6"></a>  </span>
<span id="cb92-7"><a href="#cb92-7"></a>  extracted_feats <span class="ot">=</span> feat_groups[feat_groups[, cell_type <span class="sc">==</span> ct]]<span class="sc">$</span>feature</span>
<span id="cb92-8"><a href="#cb92-8"></a>  </span>
<span id="cb92-9"><a href="#cb92-9"></a>  sign_list[[idx]] <span class="ot">=</span> extracted_feats</span>
<span id="cb92-10"><a href="#cb92-10"></a>}</span>
<span id="cb92-11"><a href="#cb92-11"></a></span>
<span id="cb92-12"><a href="#cb92-12"></a>PAGE_matrix <span class="ot">=</span> <span class="fu">makeSignMatrixPAGE</span>(<span class="at">sign_names =</span> provided_cell_types,</span>
<span id="cb92-13"><a href="#cb92-13"></a>                                 <span class="at">sign_list =</span> sign_list)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1"></a>xen_subset <span class="ot">=</span> <span class="fu">runPAGEEnrich</span>(<span class="at">gobject =</span> xen_subset,</span>
<span id="cb93-2"><a href="#cb93-2"></a>                           <span class="at">spat_unit =</span> <span class="st">"cell"</span>, </span>
<span id="cb93-3"><a href="#cb93-3"></a>                           <span class="at">feat_type =</span> <span class="st">"rna"</span>,</span>
<span id="cb93-4"><a href="#cb93-4"></a>                           <span class="at">sign_matrix =</span> PAGE_matrix,</span>
<span id="cb93-5"><a href="#cb93-5"></a>                           <span class="at">p_value =</span> <span class="cn">TRUE</span>,</span>
<span id="cb93-6"><a href="#cb93-6"></a>                           <span class="at">expression_values =</span> <span class="st">"normalized"</span>,</span>
<span id="cb93-7"><a href="#cb93-7"></a>                           <span class="at">min_overlap_genes =</span> <span class="dv">1</span>,</span>
<span id="cb93-8"><a href="#cb93-8"></a>                           <span class="at">name =</span> <span class="st">"PAGE_p_val"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Warning, Immune cells only has 1 overlapping genes. Will be removed."
[1] "Warning, Myeloid cells only has 1 overlapping genes. Will be removed."
[1] "Warning, Neutrophils only has 1 overlapping genes. Will be removed."
[1] "Warning, Plasma cells only has 1 overlapping genes. Will be removed."</code></pre>
</div>
<details close="">
<summary>Extract and Plot Enrichment Information for the StarDist Segmentation</summary>
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1"></a>p_PAGE_og <span class="ot">=</span> <span class="fu">getSpatialEnrichment</span>(xen_subset,</span>
<span id="cb95-2"><a href="#cb95-2"></a>                                 <span class="at">spat_unit =</span> <span class="st">"cell"</span>,</span>
<span id="cb95-3"><a href="#cb95-3"></a>                                 <span class="at">feat_type =</span> <span class="st">"rna"</span>, </span>
<span id="cb95-4"><a href="#cb95-4"></a>                                 <span class="at">name =</span> <span class="st">"PAGE_p_val"</span>,</span>
<span id="cb95-5"><a href="#cb95-5"></a>                                 <span class="at">output =</span> <span class="st">"data.table"</span>)</span>
<span id="cb95-6"><a href="#cb95-6"></a>n_c <span class="ot">=</span> <span class="fu">ncol</span>(p_PAGE_og)</span>
<span id="cb95-7"><a href="#cb95-7"></a></span>
<span id="cb95-8"><a href="#cb95-8"></a>p_PAGE_og[, p_max_cell_type <span class="sc">:</span><span class="er">=</span>  <span class="fu">names</span>(.SD)[<span class="fu">max.col</span>(.SD)], .SDcols <span class="ot">=</span> <span class="dv">2</span><span class="sc">:</span>n_c]</span>
<span id="cb95-9"><a href="#cb95-9"></a></span>
<span id="cb95-10"><a href="#cb95-10"></a>cID_and_types_PAGE_og_p <span class="ot">=</span> p_PAGE_og[, .(cell_ID, p_max_cell_type)]</span>
<span id="cb95-11"><a href="#cb95-11"></a>og_ct_freq <span class="ot">=</span> <span class="fu">table</span>(cID_and_types_PAGE_og_p<span class="sc">$</span>p_max_cell_type)</span>
<span id="cb95-12"><a href="#cb95-12"></a></span>
<span id="cb95-13"><a href="#cb95-13"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(cID_and_types_PAGE_og_p, <span class="fu">aes</span>(<span class="at">x =</span> p_max_cell_type)) <span class="sc">+</span> </span>
<span id="cb95-14"><a href="#cb95-14"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb95-15"><a href="#cb95-15"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>),</span>
<span id="cb95-16"><a href="#cb95-16"></a>                 <span class="at">axis.ticks.length.x =</span><span class="fu">unit</span>(<span class="fl">1.5</span>, <span class="st">"cm"</span>)) <span class="sc">+</span></span>
<span id="cb95-17"><a href="#cb95-17"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Original Cell Types from PAGE Enrichment log10(p-value)"</span>,</span>
<span id="cb95-18"><a href="#cb95-18"></a>                <span class="at">x =</span> <span class="st">"Cell Type"</span>,</span>
<span id="cb95-19"><a href="#cb95-19"></a>                <span class="at">y =</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/cell_dist_plot.png" class="img-fluid" width="672"></p>
</div>
</div>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1"></a>xen_subset <span class="ot">=</span> <span class="fu">runPAGEEnrich</span>(<span class="at">gobject =</span> xen_subset,</span>
<span id="cb96-2"><a href="#cb96-2"></a>                   <span class="at">spat_unit =</span> <span class="st">"StarDist"</span>, </span>
<span id="cb96-3"><a href="#cb96-3"></a>                   <span class="at">feat_type =</span> <span class="st">"rna"</span>,</span>
<span id="cb96-4"><a href="#cb96-4"></a>                   <span class="at">sign_matrix =</span> PAGE_matrix,</span>
<span id="cb96-5"><a href="#cb96-5"></a>                   <span class="at">p_value =</span> <span class="cn">TRUE</span>,</span>
<span id="cb96-6"><a href="#cb96-6"></a>                   <span class="at">expression_values =</span> <span class="st">"normalized"</span>,</span>
<span id="cb96-7"><a href="#cb96-7"></a>                   <span class="at">min_overlap_genes =</span> <span class="dv">1</span>,</span>
<span id="cb96-8"><a href="#cb96-8"></a>                   <span class="at">name =</span> <span class="st">"PAGE_p_val"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Warning, Immune cells only has 1 overlapping genes. Will be removed."
[1] "Warning, Myeloid cells only has 1 overlapping genes. Will be removed."
[1] "Warning, Neutrophils only has 1 overlapping genes. Will be removed."
[1] "Warning, Plasma cells only has 1 overlapping genes. Will be removed."</code></pre>
</div>
<details close="">
<summary>Extract and Plot Enrichment Information for the Original Segmentation</summary>
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1"></a>p_PAGE_sd <span class="ot">=</span> <span class="fu">getSpatialEnrichment</span>(xen_subset,</span>
<span id="cb98-2"><a href="#cb98-2"></a>                             <span class="at">spat_unit =</span> <span class="st">"StarDist"</span>,</span>
<span id="cb98-3"><a href="#cb98-3"></a>                             <span class="at">feat_type =</span> <span class="st">"rna"</span>, </span>
<span id="cb98-4"><a href="#cb98-4"></a>                             <span class="at">name =</span> <span class="st">"PAGE_p_val"</span>,</span>
<span id="cb98-5"><a href="#cb98-5"></a>                             <span class="at">output =</span> <span class="st">"data.table"</span>)</span>
<span id="cb98-6"><a href="#cb98-6"></a>n_c <span class="ot">=</span> <span class="fu">ncol</span>(p_PAGE_sd)</span>
<span id="cb98-7"><a href="#cb98-7"></a></span>
<span id="cb98-8"><a href="#cb98-8"></a>p_PAGE_sd[, p_max_cell_type <span class="sc">:</span><span class="er">=</span>  <span class="fu">names</span>(.SD)[<span class="fu">max.col</span>(.SD)], .SDcols <span class="ot">=</span> <span class="dv">2</span><span class="sc">:</span>n_c]</span>
<span id="cb98-9"><a href="#cb98-9"></a></span>
<span id="cb98-10"><a href="#cb98-10"></a>cID_and_types_PAGE_sd_p <span class="ot">=</span> p_PAGE_sd[, .(cell_ID, p_max_cell_type)]</span>
<span id="cb98-11"><a href="#cb98-11"></a>sd_ct_freq <span class="ot">=</span> <span class="fu">table</span>(cID_and_types_PAGE_sd_p<span class="sc">$</span>p_max_cell_type)</span>
<span id="cb98-12"><a href="#cb98-12"></a></span>
<span id="cb98-13"><a href="#cb98-13"></a>ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(cID_and_types_PAGE_sd_p, <span class="fu">aes</span>(<span class="at">x =</span> p_max_cell_type)) <span class="sc">+</span> </span>
<span id="cb98-14"><a href="#cb98-14"></a>  ggplot2<span class="sc">::</span><span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb98-15"><a href="#cb98-15"></a>  ggplot2<span class="sc">::</span><span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>),</span>
<span id="cb98-16"><a href="#cb98-16"></a>                 <span class="at">axis.ticks.length.x =</span><span class="fu">unit</span>(<span class="fl">1.5</span>, <span class="st">"cm"</span>)) <span class="sc">+</span></span>
<span id="cb98-17"><a href="#cb98-17"></a>  ggplot2<span class="sc">::</span><span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"StarDist Cell Types from PAGE Enrichment log10(p-value)"</span>,</span>
<span id="cb98-18"><a href="#cb98-18"></a>                <span class="at">x =</span> <span class="st">"Cell Type"</span>,</span>
<span id="cb98-19"><a href="#cb98-19"></a>                <span class="at">y =</span> <span class="st">"Frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/sd_cell_dist_plot.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="plot-cell-type-distributions" class="level3" data-number="9.2">
<h3 data-number="9.2" class="anchored" data-anchor-id="plot-cell-type-distributions"><span class="header-section-number">9.2</span> Plot Cell Type Distributions</h3>
<div class="cell">
<details close="">
<summary>Compare Cell Typing for Each Segmentation</summary>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1"></a>my_colors <span class="ot">=</span> <span class="fu">getDistinctColors</span>(<span class="dv">16</span>)</span>
<span id="cb99-2"><a href="#cb99-2"></a></span>
<span id="cb99-3"><a href="#cb99-3"></a>og_cell_types <span class="ot">=</span> <span class="fu">unique</span>(cID_and_types_PAGE_og_p<span class="sc">$</span>p_max_cell_type)</span>
<span id="cb99-4"><a href="#cb99-4"></a>og_ct_freq_dt <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">data.table</span>(og_ct_freq)</span>
<span id="cb99-5"><a href="#cb99-5"></a><span class="fu">colnames</span>(og_ct_freq_dt) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"num_cells"</span>)</span>
<span id="cb99-6"><a href="#cb99-6"></a></span>
<span id="cb99-7"><a href="#cb99-7"></a>og_total_cells <span class="ot">=</span> <span class="fu">length</span>(cID_and_types_PAGE_og_p<span class="sc">$</span>p_max_cell_type)</span>
<span id="cb99-8"><a href="#cb99-8"></a></span>
<span id="cb99-9"><a href="#cb99-9"></a><span class="cf">for</span> ( i <span class="cf">in</span> og_cell_types){</span>
<span id="cb99-10"><a href="#cb99-10"></a>  nullvar <span class="ot">=</span> og_ct_freq_dt[cell_type <span class="sc">==</span> i, perc <span class="sc">:</span><span class="er">=</span> num_cells<span class="sc">/</span><span class="fu">sum</span>(og_ct_freq_dt<span class="sc">$</span>num_cells) <span class="sc">*</span> <span class="dv">100</span>]</span>
<span id="cb99-11"><a href="#cb99-11"></a>}</span>
<span id="cb99-12"><a href="#cb99-12"></a></span>
<span id="cb99-13"><a href="#cb99-13"></a>pl_og <span class="ot">=</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(og_ct_freq_dt), <span class="fu">aes</span>(<span class="at">x=</span><span class="st">""</span>, <span class="at">y=</span>perc, <span class="at">fill =</span> cell_type)) <span class="sc">+</span></span>
<span id="cb99-14"><a href="#cb99-14"></a>          <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">width =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb99-15"><a href="#cb99-15"></a>          <span class="fu">coord_polar</span>(<span class="st">"y"</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb99-16"><a href="#cb99-16"></a>          <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb99-17"><a href="#cb99-17"></a>          <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb99-18"><a href="#cb99-18"></a>          <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">"Original Cell Types ("</span>, <span class="fu">as.character</span>(og_total_cells), <span class="st">" Cells)"</span>))</span>
<span id="cb99-19"><a href="#cb99-19"></a></span>
<span id="cb99-20"><a href="#cb99-20"></a></span>
<span id="cb99-21"><a href="#cb99-21"></a>sd_cell_types <span class="ot">=</span> <span class="fu">unique</span>(cID_and_types_PAGE_sd_p<span class="sc">$</span>p_max_cell_type)</span>
<span id="cb99-22"><a href="#cb99-22"></a>sd_ct_freq_dt <span class="ot">=</span> data.table<span class="sc">::</span><span class="fu">data.table</span>(sd_ct_freq)</span>
<span id="cb99-23"><a href="#cb99-23"></a><span class="fu">colnames</span>(sd_ct_freq_dt) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"cell_type"</span>, <span class="st">"num_cells"</span>)</span>
<span id="cb99-24"><a href="#cb99-24"></a></span>
<span id="cb99-25"><a href="#cb99-25"></a>sd_total_cells <span class="ot">=</span> <span class="fu">length</span>(cID_and_types_PAGE_sd_p<span class="sc">$</span>p_max_cell_type)</span>
<span id="cb99-26"><a href="#cb99-26"></a></span>
<span id="cb99-27"><a href="#cb99-27"></a><span class="cf">for</span> ( i <span class="cf">in</span> sd_cell_types){</span>
<span id="cb99-28"><a href="#cb99-28"></a>  nullvar <span class="ot">=</span> sd_ct_freq_dt[cell_type <span class="sc">==</span> i, perc <span class="sc">:</span><span class="er">=</span> num_cells<span class="sc">/</span><span class="fu">sum</span>(sd_ct_freq_dt<span class="sc">$</span>num_cells) <span class="sc">*</span> <span class="dv">100</span>]</span>
<span id="cb99-29"><a href="#cb99-29"></a>}</span>
<span id="cb99-30"><a href="#cb99-30"></a></span>
<span id="cb99-31"><a href="#cb99-31"></a>pl_sd <span class="ot">=</span> ggplot2<span class="sc">::</span><span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(sd_ct_freq_dt), <span class="fu">aes</span>(<span class="at">x=</span><span class="st">""</span>, <span class="at">y=</span>perc, <span class="at">fill =</span> cell_type)) <span class="sc">+</span></span>
<span id="cb99-32"><a href="#cb99-32"></a>          <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>, <span class="at">width =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb99-33"><a href="#cb99-33"></a>          <span class="fu">coord_polar</span>(<span class="st">"y"</span>, <span class="at">start =</span> <span class="dv">0</span>) <span class="sc">+</span></span>
<span id="cb99-34"><a href="#cb99-34"></a>          <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb99-35"><a href="#cb99-35"></a>          <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb99-36"><a href="#cb99-36"></a>          <span class="fu">labs</span>(<span class="at">title =</span> <span class="fu">paste</span>(<span class="st">"StarDist Cell Types ("</span>, <span class="fu">as.character</span>(sd_total_cells), <span class="st">" Cells)"</span>))</span>
<span id="cb99-37"><a href="#cb99-37"></a></span>
<span id="cb99-38"><a href="#cb99-38"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(pl_sd, pl_og, <span class="at">align =</span> <span class="st">"h"</span>, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="xenium_MOB_files/figure-html/pie.png" class="img-fluid" width="802"></p>
</div>
</div>
</section>
</section>
<section id="save-the-giotto-objects" class="level2" data-number="10">
<h2 data-number="10" class="anchored" data-anchor-id="save-the-giotto-objects"><span class="header-section-number">10</span> Save the Giotto Objects</h2>
<p>To save the Giotto Objects, provide an existing subdirectory within <code>results_folder</code> to the <code>foldername</code> argument. These objects may be loaded calling <code>loadGiotto("/path/to/foldername/")</code>.</p>
<div class="cell">
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1"></a><span class="fu">saveGiotto</span>(<span class="at">gobject =</span> xenium_gobj,</span>
<span id="cb100-2"><a href="#cb100-2"></a>           <span class="at">foldername =</span>  <span class="st">'saved_gobjects'</span>,</span>
<span id="cb100-3"><a href="#cb100-3"></a>           <span class="at">dir =</span> results_folder,</span>
<span id="cb100-4"><a href="#cb100-4"></a>           <span class="at">image_filetype =</span> <span class="st">'.tif'</span>,</span>
<span id="cb100-5"><a href="#cb100-5"></a>           <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Folder already exist and overwrite = TRUE, overwrite folder</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>1. Start writing feature information</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: rna</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: blank_code</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: neg_code</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: neg_probe</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2. Start writing spatial information</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For spatial information: cell</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For spatial information: nucleus</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>3. Start writing image information</code></pre>
</div>
<details open="">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1"></a><span class="fu">saveGiotto</span>(<span class="at">gobject =</span> xen_subset,</span>
<span id="cb111-2"><a href="#cb111-2"></a>           <span class="at">foldername =</span>  <span class="st">'saved_gobjects_subset'</span>,</span>
<span id="cb111-3"><a href="#cb111-3"></a>           <span class="at">dir =</span> results_folder,</span>
<span id="cb111-4"><a href="#cb111-4"></a>           <span class="at">image_filetype =</span> <span class="st">'.tif'</span>,</span>
<span id="cb111-5"><a href="#cb111-5"></a>           <span class="at">overwrite =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Folder already exist and overwrite = TRUE, overwrite folder</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>1. Start writing feature information</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: rna</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: blank_code</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: neg_code</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For feature: neg_probe</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>2. Start writing spatial information</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For spatial information: cell</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For spatial information: nucleus</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>For spatial information: StarDist</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>3. Start writing image information</code></pre>
</div>
</div>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>